File: model/src/main/java/io/github/runtimemodels/chazm/relation/NeedsFactory.java
package io.github.runtimemodels.chazm.relation;

import io.github.runtimemodels.chazm.entity.Attribute;
import io.github.runtimemodels.chazm.entity.Role;
import io.github.runtimemodels.chazm.relation.Needs;

/**
 * The {@linkplain NeedsFactory} interface defines the API for constructing {@linkplain Needs} relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface NeedsFactory {

	/**
	 * Constructs a {@linkplain Needs}.
	 *
	 * @param role
	 *            the {@linkplain Role} of the {@linkplain Needs}.
	 * @param attribute
	 *            the {@linkplain Attribute} of the {@linkplain Needs}.
	 * @return a {@linkplain Needs}.
	 */
	Needs buildNeeds(Role role, Attribute attribute);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/factory/relation/NeedsFactory.kt
package runtimemodels.chazm.model.factory.relation

import runtimemodels.chazm.api.entity.Attribute
import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.relation.Needs

/**
 * The [NeedsFactory] interface defines the API for constructing [Needs] relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface NeedsFactory {
    /**
     * Constructs a [Needs].
     *
     * @param role      the [Role] of the [Needs].
     * @param attribute the [Attribute] of the [Needs].
     * @return a [Needs].
     */
    fun build(role: Role, attribute: Attribute): Needs
}
